File: Sean的leetcode/Leetcode/Moving Average From Data Stream.cpp
//Use queue for FIFO 
class MovingAverage {
public:
    /** Initialize your data structure here. */
    MovingAverage(int size):size(size),sum(0) {
        
    }
    
    double next(int val) {
        if (q.size() == maxsize) {
            sum -= q.front();
            q.pop();
        }
        q.push(val);
        sum += val;
        return sum / nums.size();
    }
private:
    double sum;
    int size;
    queue<int> nums;
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */

File: Sean的leetcode/python/Moving Average from Data Stream.py
###
# Queue/Deque
# Time Complexity: O(1) 
# Space Complexity: O(size)
###
class MovingAverage(object):

    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        from collections import deque
        self.nums = deque()
        self.size = size
        self.sum = 0

    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        # move left end
        if len(self.nums) == self.size:
            self.sum -= self.nums.popleft()
        # insert element
        self.nums.append(val)
        self.sum += val
        return float(self.sum) / len(self.nums)


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)
        
        